### PR TITLE
feat: display cycle count on CI

### PR DESCRIPTION
**Pull request by leruaa**
The goal of this PR is to display stats on a proof execution on CI.

In order to not pollute `multi` binary with CI related stuff, I did a small refactoring of the witness generation and proof execution part, moving them to `lib` so they can be used by the binary and also an integration test.

The code that call Github API is mainly copied from the one in SP1.

I also did a small change to the cost estimator binary to alway take a block range in the last 2 weeks.

### CODE REVIEW

File: scripts/prove/bin/multi.rs [hunk 1]
```diff
@@ -113,37 +100,32 @@ async fn main() -> Result<()> {
...
-
-        // Create the report directory if it doesn't exist.
-        let report_dir = format!("execution-reports/multi/{}", l2_chain_id);
-        if !std::path::Path::new(&report_dir).exists() {
-            fs::create_dir_all(&report_dir).unwrap();
-        }
-
-        let block_data = data_fetcher
-            .get_l2_block_data_range(l2_start_block, l2_end_block)
-            .await?;
```
Comment (ratankaliani) [inline scripts/prove/bin/multi.rs:137]: Why did you remove this? We need to create the directory unless it already exists here.

Comment (ratankaliani) [reply scripts/prove/bin/multi.rs:137]: nvm

File: utils/host/src/block_range.rs [hunk 2]
```diff
@@ -42,3 +42,17 @@ pub async fn get_validated_block_range(
 
     Ok((l2_start_block, l2_end_block))
 }
+
+// Get a fixed recent (less than 2 weeks) block range.
```
Comment (ratankaliani) [inline utils/host/src/block_range.rs:46]: nit: Make a Rustdoc

File: utils/host/src/block_range.rs [hunk 3]
```diff
@@ -42,3 +42,17 @@ pub async fn get_validated_block_range(
...
+pub async fn get_rolling_block_range(
+    data_fetcher: &OPSuccinctDataFetcher,
+    range: u64,
+) -> Result<(u64, u64)> {
+    let header = data_fetcher.get_l2_header(BlockId::finalized()).await?;
+    let l2_block_time = data_fetcher.get_l2_block_time().await?;
+    let block_count_in_two_weeks = 14 * 24 * 60 * 60 / l2_block_time;
+    let l2_end_block = (header.number / block_count_in_two_weeks) * block_count_in_two_weeks;
```
Comment (ratankaliani) [inline utils/host/src/block_range.rs:54]: Why not just use `header.number` here? You're diving by, then multiply by `block_count_in_two_weeks`.

File: utils/host/src/block_range.rs (hunk 2 above)
Comment (ratankaliani) [inline utils/host/src/block_range.rs:46]: Use rustdoc `///` for fn comments.

File: scripts/prove/tests/common/mod.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
+use reqwest::Client;
+use serde_json::json;
+
+pub async fn post_to_github_pr(
```
Comment (ratankaliani) [inline scripts/prove/tests/common/mod.rs:4]: nit: Add fn comment

File: scripts/prove/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    let witness_generation_time_sec = start_time.elapsed();
+    println!(
+        "Witness Generation Duration: {:?}",
+        witness_generation_time_sec.as_secs()
+    );
+
+    Ok(witness_generation_time_sec)
+}
+
+pub async fn execute_proof(
```
Comment (ratankaliani) [inline scripts/prove/src/lib.rs:35]: `execute_multi` is a better name

File: scripts/prove/tests/multi.rs
[the remark this answers is not in the document]
Comment (ratankaliani) [reply scripts/prove/tests/multi.rs:1]: nvm, probably good given that it's CI

[the remark this answers is not in the document]
Comment (leruaa) [reply scripts/prove/tests/multi.rs:1]: There are a few reasons:

- I didn't wanted to "pollute" the binary with CI stuff (i.e. sending to GitHub API)
- The function that create the comment in GitHub can easily used in other integration tests
- If we add other integration tests, we could run them in paralell
- This IS an integration test ;)

File: utils/host/src/block_range.rs [hunk 6]
```diff
@@ -42,3 +44,17 @@ pub async fn get_validated_block_range(
...
     Ok((l2_start_block, l2_end_block))
 }
+
+/// Get a fixed recent (less than 2 weeks) block range.
+pub async fn get_rolling_block_range(
+    data_fetcher: &OPSuccinctDataFetcher,
+    range: u64,
+) -> Result<(u64, u64)> {
+    let now = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs();
+    let start_timestanp = now - (now % (14 * 24 * 60 * 60));
```
Comment (ratankaliani) [inline utils/host/src/block_range.rs:54]: nit: Pass the "fixed" window (2 weeks) to the function, and use a constant for it inside of the multi/cost estimator script.

File: scripts/prove/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        .execute(MULTI_BLOCK_ELF, sp1_stdin.clone())
+        .run()
+        .unwrap();
+    let execution_duration = start_time.elapsed();
+
+    // Create the report directory if it doesn't exist.
+    let report_dir = format!("execution-reports/multi/{}", l2_chain_id);
+    if !std::path::Path::new(&report_dir).exists() {
+        fs::create_dir_all(&report_dir)?;
+    }
```
Comment (ratankaliani) [inline scripts/prove/src/lib.rs:54]: Taste, but this shouldn't occur inside of this function, should instead be inside of the binary/test that's running it, as you materialize the `report_path` separately there as well. Would remove this and move it to the `multi` script.

**Review by ratankaliani — APPROVED** [review]
Small refactoring/naming convention changes, otherwise LGTM
